[baremetal] Add random delay function to platform utils

artokin opened the pull request:
## Description

Add delay function `mbedtls_platform_random_delay` to platform_utils. The function will delay
program execution by incrementing local variable and doing xor byte operations a randomised number of times.

This PR contains an API break, function `mbedtls_platform_enforce_volatile_reads` is removed. The same kind of behaviour is achieved by using function `mbedtls_platform_random_delay(0)`.

__Impact on code-size:__

| | ARMC5 |
| --- | --- |
| `libmbedcrypto.a` before | 18659 |
| `libmbedcrypto.a` after | 18699 |
| gain in bytes Crypto | -40 |


## Status
**READY**

jarvte commented on include/mbedtls/platform_util.h:
```diff
@@ -238,6 +238,21 @@ int mbedtls_platform_memcmp( const void *buf1, const void *buf2, size_t num );
+ *
+ * \return      In success number of increments, -1 in case of errors.
+ */
+int mbedtls_platform_random_delay( size_t num );
```
Should there be a default value for num? Or at least some hint how much delay does this generate with different num values?

artokin replied:
I tried to rephrase the function comments. 
The purpose of this function is to generate a random delay, number of variable increments can be anything between 1 to `num`. I feel that "default value" is not suitable for function that is random by nature.

Function caller should not use this function to create a predictable delay. There should be other, more sophisticated methods for generating deterministic delay.

jarlamsa commented on library/platform_util.c:
```diff
@@ -168,13 +168,34 @@ uint32_t mbedtls_platform_random_in_range( size_t num )
 void mbedtls_platform_enforce_volatile_reads( void )
 {
+    // Add a small random delay as a counter-measure to fault injection attack.
+    mbedtls_platform_random_delay( 50 );
```
Not sure if it is a good idea to use magic number here instead of defining the constant somewhere.

artokin replied:
I was afraid that the constant would be used as a default delay for all places where `mbedtls_platform_random_delay` gets called. Therefore used magic number here.

artokin commented:
With two approving reviews and passing CI setting status to "ready for merge".

sbutcher-arm commented on library/platform_util.c:
```diff
@@ -168,13 +168,34 @@ uint32_t mbedtls_platform_random_in_range( size_t num )
+
+    do
+    {
+        i++;
```
Major: We should do some random work here, not just increment a counter. In a power trace, it probably looks like a uniform delay, and may easily be removed from the trace by analysis.

We should extract two random `uint32` words, and xor themselves together repeatedly. That will show variation in the power trace on each iteration of the loop.

artokin replied:
Thanks @sbutcher-arm , this is fixed in the commit "_Add more variation to random delay countermeasure_"

sbutcher-arm commented on library/platform_util.c:
```diff
@@ -168,13 +168,34 @@ uint32_t mbedtls_platform_random_in_range( size_t num )
+#endif /* !MBEDTLS_ENTROPY_HARDWARE_ALT */
+}
+
 void mbedtls_platform_enforce_volatile_reads( void )
```
Minor: The function name made sense when it was a dummy function, but now it's actually doing something every time in providing a random delay. 

Given it's simply calling a function of `mbedtls_platform_random_delay`, I'm not sure we need both functions, and together they don't make sense.

artokin commented:
Review comments addressed, @jarlamsa and @sbutcher-arm would you please re-review?

jarlamsa commented:
This now causes an API break as the `mbedtls_platform_enforce_volatile_reads` gets removed, it should be noted in the description of this PR.

sbutcher-arm commented:
Re-running the CI.

artokin commented:
Rebased and replaced new occurrences of `mbedtls_platform_enforce_volatile_reads` with `mbedtls_platform_random_delay`.

@jarlamsa , @jarvte  and @sbutcher-arm would you please re-review?

artokin commented:
rebased, please re-review.

artokin commented:
With two approving reviews and CI indicating the intentional API break setting status to "ready for merge".

sbutcher-arm commented on library/platform_util.c:
```diff
@@ -165,13 +165,37 @@ uint32_t mbedtls_platform_random_in_range( size_t num )
+    size_t rn_1, rn_2, rn_3;
+    volatile size_t i = 0;
+    uint8_t shift;
+    if( max_rand == 0 || max_rand > INT_MAX )
```
Major - I think this check is vulnerable to fault injection attack. You can force an early return of the function.

The problem here - is if you do a first fault injection, this check will be a *fixed* period after the first, and allow you to bypass the secondary check.

Sorry, but this is flawed.

jarlamsa replied:
One thing to note is that this is used between double checks in other code, so potential attacker would need 3 consecutive faults to be injected to possibly benefit from skipping the random delay. Anyway, this check would probably more preferably just use some default value for the max_rand instead of returning -1. (As the return values aren't checked anywhere)

sbutcher-arm commented on library/platform_util.c:
```diff
@@ -165,13 +165,37 @@ uint32_t mbedtls_platform_random_in_range( size_t num )
- * them. This functions doesn't need to do anything, it just needs to be
- * in another compilation unit. So here's a function that does nothing. */
-void mbedtls_platform_enforce_volatile_reads( void )
+int mbedtls_platform_random_delay( size_t max_rand )
```
Suggestion - I don't think we need to parameterise this function. Normally, adding a parameter to a function makes better more flexible code, but here, it creates a new angle for potential vulnerabilities.

In real world usage, there's no reason we couldn't make this a compile time option, and fix what the maximum delay is. I don't think there's any reason why we'd vary it per call, nor at runtime.

Therefore I suggest you remove `max_rand` and replace it with a constant `MBEDTLS_MAX_RAND_DELAY`.

artokin replied:
mbedtls_platform_random_delay signature changed in the next commit, this also saves 44 bytes of ROM.

artokin commented:
Review comments addressed, would you please reX-review :-)

artokin commented:
3 approvals and CI failing to API check as expected, promoting to `ready for merge`.
